### ISSUE TITLE
Does not converge sufficiently

### ISSUE THREAD

**Issue by NicolasPPX** [mori97/U-Net_MUSDB18#2]
Hi i have coded your network but the convergence isn't sufficiently good enough and the network converge to the input image (mix) not the vocal one. This is the network and some code that drives the network : 

class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = torch.nn.Conv2d(1, 16, kernel_size=(5,5), stride=(2, 2), padding=2)
        self.conv_bn1 = torch.nn.BatchNorm2d(16)
        self.conv2 = torch.nn.Conv2d(16, 32, kernel_size=(5,5), stride=(2, 2), padding=2)
        self.conv_bn2 = torch.nn.BatchNorm2d(32)
        self.conv3 = torch.nn.Conv2d(32, 64, kernel_size=(5,5), stride=(2, 2), padding=2)
        self.conv_bn3 = torch.nn.BatchNorm2d(64)
        self.conv4 = torch.nn.Conv2d(64, 128, kernel_size=(5,5), stride=(2, 2), padding=2)
        self.conv_bn4 = torch.nn.BatchNorm2d(128)
        self.conv5 = torch.nn.Conv2d(128, 256, kernel_size=(5,5), stride=(2, 2), padding=2)
        self.conv_bn5 = torch.nn.BatchNorm2d(256)
        self.conv6 = torch.nn.Conv2d(256, 512, kernel_size=(5,5), stride=(2, 2), padding=2)
        self.conv_bn6 = torch.nn.BatchNorm2d(512)
      
        self.deconv1 = torch.nn.ConvTranspose2d(512, 256, kernel_size=(5,5), stride=(2, 2), padding=2)
        self.deconv_bn1 = torch.nn.BatchNorm2d(256)
        self.dropout1 = torch.nn.Dropout2d(0.5)
        self.deconv2 = torch.nn.ConvTranspose2d(512, 128, kernel_size=(5,5), stride=(2, 2), padding=2)
        self.deconv_bn2 = torch.nn.BatchNorm2d(128)
        self.dropout2 = torch.nn.Dropout2d(0.5)
        self.deconv3 = torch.nn.ConvTranspose2d(256, 64, kernel_size=(5,5), stride=(2, 2), padding=2)
        self.deconv_bn3 = torch.nn.BatchNorm2d(64)
        self.dropout3 = torch.nn.Dropout2d(0.5)
        self.deconv4 = torch.nn.ConvTranspose2d(128, 32, kernel_size=(5,5), stride=(2, 2), padding=2)
        self.deconv_bn4 = torch.nn.BatchNorm2d(32)
        self.deconv5 = torch.nn.ConvTranspose2d(64, 16, kernel_size=(5,5), stride=(2, 2), padding=2)
        self.deconv_bn5 = torch.nn.BatchNorm2d(16)
        self.deconv6 = torch.nn.ConvTranspose2d(32, 1, kernel_size=(5,5), stride=(2, 2), padding=2)

    def forward(self, x):
        
        h1 = F.leaky_relu(self.conv_bn1(self.conv1(x)))
        h2 = F.leaky_relu(self.conv_bn2(self.conv2(h1)))
        h3 = F.leaky_relu(self.conv_bn3(self.conv3(h2)))
        h4 = F.leaky_relu(self.conv_bn4(self.conv4(h3)))
        h5 = F.leaky_relu(self.conv_bn5(self.conv5(h4)))
        h = F.leaky_relu(self.conv_bn6(self.conv6(h5)))

        h = self.dropout1(F.relu(self.deconv_bn1(self.deconv1(h, output_size = h5.size()))))
        h = torch.cat((h, h5), dim=1)
        h = self.dropout2(F.relu(self.deconv_bn2(self.deconv2(h, output_size = h4.size()))))
        h = torch.cat((h, h4), dim=1)
        h = self.dropout3(F.relu(self.deconv_bn3(self.deconv3(h, output_size = h3.size()))))
        h = torch.cat((h, h3), dim=1)
        h = F.relu(self.deconv_bn4(self.deconv4(h, output_size = h2.size())))
        h = torch.cat((h, h2), dim=1)
        h = F.relu(self.deconv_bn5(self.deconv5(h, output_size = h1.size())))
        h = torch.cat((h, h1), dim=1)
        h = torch.sigmoid(self.deconv6(h))
        return h

n_files = len(mix_all[0,0,:])

net = Net()
optimizer = torch.optim.Adam(net.parameters(), lr=0.001)
criterion = nn.MSELoss()

loss_values = []
for epoch in range(0,100,1):
    for i in range(0,n_files,1):
        inputs = mix_all[:,:,i]   
        outputs = vocal_all[:,:,i]
               
        optimizer.zero_grad()
    
        inputs = torch.tensor(inputs) 
        inputs = torch.unsqueeze(inputs,0)
        inputs = torch.unsqueeze(inputs,0)

        outputs = torch.tensor(outputs) 
        outputs_1d = outputs.flatten()
        outputs_1d = outputs_1d.float()
    
        outputs_net = net(inputs.float())
        outputs_net = outputs_net * inputs.float()
        outputs_net = torch.squeeze(outputs_net,0)
        outputs_net = torch.squeeze(outputs_net,0)
        outputs_net_1d = outputs_net.flatten()
    
        loss = criterion(outputs_1d, outputs_net_1d)
        #loss = F.l1_loss(outputs, outputs_net, reduction='sum')
    
        print('epoch %i, %.3f / 1 : %.8f' %(epoch,i/n_files,loss))
    
        loss_values.append(loss)
        
        loss.backward()
        optimizer.step()

Many thanks.
